File: lemonad-matlab/day9/day9.m
%
% Day 9, Advent of code 2017 (Jonas Nockert / @lemonad)
%

runtests('day9Test.m');

[s, g] = solve();
fprintf("Total score for all groups in input: %d\n", s);
fprintf("Number of non-canceled characters within garbage in input?: %d\n", g);
assert(s == 9251)
assert(g == 4322)


%
% Helpers
%

function [score, garbage_count] = solve()
    score = 0;
    garbage_count = 0;
    fp = fopen('day9.in');
    line = fgetl(fp);
    while ischar(line)
        [subscore, count] = parseSequence(line);
        score = score + subscore;
        garbage_count = garbage_count + count;
        line = fgetl(fp);
    end
end

File: meldanya-python3/day9/day9.py
def solve(line):
    sum_, garbage, n = 0, 0, 0
    ignore = False
    prev = None

    for ch in line:
        if (ignore and ch != '>' and ch != '!' and prev != '!'):
            garbage += 1

        if ch == '{' and not ignore and prev != '!':
            n += 1
            sum_ += n
        elif ch == '}' and not ignore and prev != '!':
            n -= 1
        elif ch == '<' and prev != '!':
            ignore = True
        elif ch == '>' and prev != '!':
            ignore = False

        prev = ch if ch != '!' or prev != '!' else None

    return sum_, garbage


def main():
    line = input()
    print(solve(line))


if __name__ == '__main__':
    main()
